ebms-send-in/src/main/kotlin/no/nav/emottak/ebms/App.kt

```diff
@@ -1,199 +1,60 @@
 import io.ktor.server.application.Application
-import io.ktor.server.application.install
-import io.ktor.server.auth.Authentication
-import io.ktor.server.auth.authenticate
-import io.ktor.server.engine.embeddedServer
-import io.ktor.server.metrics.micrometer.MicrometerMetrics
 import io.ktor.server.netty.Netty
-import io.ktor.server.plugins.contentnegotiation.ContentNegotiation
-import io.ktor.server.request.receive
-import io.ktor.server.response.respond
-import io.ktor.server.routing.post
-import io.ktor.server.routing.routing
-import io.micrometer.core.instrument.Timer.ResourceSample
+import io.ktor.utils.io.CancellationException
 import io.micrometer.prometheus.PrometheusConfig
 import io.micrometer.prometheus.PrometheusMeterRegistry
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.withContext
-import kotlinx.serialization.json.Json
-import no.nav.emottak.auth.AZURE_AD_AUTH
-import no.nav.emottak.auth.AuthConfig
-import no.nav.emottak.ebms.SupportedService.Companion.toSupportedService
-import no.nav.emottak.fellesformat.FellesFormatXmlMarshaller
-import no.nav.emottak.fellesformat.asEIFellesFormat
-import no.nav.emottak.frikort.frikortsporring
-import no.nav.emottak.frikort.frikortsporringMengde
-import no.nav.emottak.melding.model.SendInRequest
-import no.nav.emottak.melding.model.SendInResponse
-import no.nav.emottak.pasientliste.PasientlisteService
-import no.nav.emottak.utbetaling.UtbetalingClient
-import no.nav.emottak.utbetaling.UtbetalingXmlMarshaller
-import no.nav.emottak.util.LogLevel
-import no.nav.emottak.util.asJson
-import no.nav.emottak.util.asXml
-import no.nav.emottak.util.marker
-import no.nav.emottak.utils.isProdEnv
-import no.nav.security.token.support.v3.tokenValidationSupport
+import kotlinx.coroutines.awaitCancellation
+import no.nav.emottak.config.config
+import no.nav.emottak.ebms.plugin.configureAuthentication
+import no.nav.emottak.ebms.plugin.configureContentNegotiation
+import no.nav.emottak.ebms.plugin.configureCoroutineDebugger
+import no.nav.emottak.ebms.plugin.configureMetrics
+import no.nav.emottak.ebms.plugin.configureRoutes
 import org.slf4j.LoggerFactory
-import kotlin.uuid.ExperimentalUuidApi
-import kotlin.uuid.Uuid
 
 internal val log = LoggerFactory.getLogger("no.nav.emottak.ebms.App")
 
-fun main() {
-    log.info("ebms-send-in starting")
-
-    System.setProperty("io.ktor.http.content.multipart.skipTempFile", "true")
-    embeddedServer(Netty, port = 8080, module = Application::ebmsSendInModule)
-        .also { it.engineConfig.maxChunkSize = 100000 }
-        .start(wait = true)
-}
-
-fun <T> timed(meterRegistry: PrometheusMeterRegistry, metricName: String, process: ResourceSample.() -> T): T =
-    io.micrometer.core.instrument.Timer.resource(meterRegistry, metricName)
-        .use {
-            process(it)
+fun main() = SuspendApp {
+    result {
+        resourceScope {
+            setupServer()
+            awaitCancellation()
+        }
+    }.onFailure { error ->
```

Comment:
Sikkert inspirert av det jeg har gjort, men denne kan godt slankes litt:

`.onFailure { error -> if (error !is CancellationException) logError(error) }`

Reply:
Veldig bra forslag :+1: